src/server.rs

```diff
@@ -335,80 +329,135 @@ fn collect_changes(
+    for (init_message, update_message, client_info) in messages.iter_mut_with_info() {
+        if ticks.is_added(change_tick.last_run(), change_tick.this_run()) {
+            init_message.write_component(replication_info, replication_id, component)?;
+        } else {
+            let tick = *client_info
+                .ticks
+                .get(&entity)
+                .expect("entity should present after adding component");
+            if ticks.is_changed(tick, change_tick.this_run()) {
+                update_message.write_component(replication_info, replication_id, component)?;
+            }
+        }
     }
 
     Ok(())
 }
 
-/// Collect component removals into messages based on last acknowledged tick.
+/// Collects component removals from this tick into init messages.
 fn collect_removals(
-    messages: &mut [ReplicationMessage],
-    removal_trackers: &Query<(Entity, &RemovalTracker)>,
-    system_tick: Tick,
+    messages: &mut ReplicationMessages,
+    remove_events: &RemovedComponentEvents,
+    tick: Tick,
+    replication_rules: &ReplicationRules,
 ) -> bincode::Result<()> {
-    for message in &mut *messages {
+    for (message, _) in messages.iter_mut() {
         message.start_array();
     }
 
-    for (entity, removal_tracker) in removal_trackers {
-        for message in &mut *messages {
+    // PERF: Unfortunately, removed components are grouped by type, not by entity.
+    // This is why we need an intermediate container. But in practice users rarely
+    // remove a lot of components in the same tick, so it's probably fine.
+    let mut removals: EntityHashMap<_, Vec<_>> = Default::default();
+    for (&component_id, &replication_id) in replication_rules.get_ids() {
+        for entity in remove_events
+            .get(component_id)
+            .into_iter()
+            .flat_map(|removed| removed.iter_current_update_events().cloned())
+            .map(Into::into)
+        {
+            removals.entry(entity).or_default().push(replication_id);
+        }
+    }
+
+    for (entity, components) in removals {
+        for (message, _, client_info) in messages.iter_mut_with_info() {
             message.start_entity_data(entity);
-            for (&replication_id, &tick) in &removal_tracker.0 {
-                if tick.is_newer_than(message.system_tick, system_tick) {
-                    message.write_replication_id(replication_id)?;
-                }
+            for &replication_id in &components {
+                client_info.ticks.insert(entity, tick);
+                message.write_replication_id(replication_id)?;
             }
             message.end_entity_data()?;
         }
     }
 
-    for message in &mut *messages {
+    for (message, _) in messages.iter_mut() {
         message.end_array()?;
     }
 
     Ok(())
 }
 
-/// Collect entity despawns into messages based on last acknowledged tick.
+/// Collect entity despawns from this tick into init messages.
 fn collect_despawns(
```

Comment:
I'm not quite sure about this... Yes, this may be more cache-friendly, but I would like to avoid complicating the code without benchmarking cases like this first.

Also I suspect that it won't be needed if we write everything into a single buffer and remember only positions for clients (the thing you suggested in Discord, planning to play with it ASAP).

Reply:
Opened: https://github.com/lifescapegame/bevy_replicon/issues/129